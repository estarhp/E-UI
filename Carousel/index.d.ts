import React from "react";
import "./style/index.scss";
export interface CarouselProps {
    className?: string;
    children: React.ReactNode;
    time?: number;
}
declare const Carousel: React.ForwardRefExoticComponent<CarouselProps & React.RefAttributes<HTMLDivElement>>;
export default Carousel;
