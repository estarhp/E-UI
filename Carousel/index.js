var _excluded = ["className", "children", "time"];
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = null == arr ? null : "undefined" != typeof Symbol && arr[Symbol.iterator] || arr["@@iterator"]; if (null != _i) { var _s, _e, _x, _r, _arr = [], _n = !0, _d = !1; try { if (_x = (_i = _i.call(arr)).next, 0 === i) { if (Object(_i) !== _i) return; _n = !1; } else for (; !(_n = (_s = _x.call(_i)).done) && (_arr.push(_s.value), _arr.length !== i); _n = !0); } catch (err) { _d = !0, _e = err; } finally { try { if (!_n && null != _i.return && (_r = _i.return(), Object(_r) !== _r)) return; } finally { if (_d) throw _e; } } return _arr; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
import React, { forwardRef, useEffect, useState } from "react";
import classNames from "classnames";
import "./style/index.scss";
var Carousel = /*#__PURE__*/forwardRef(function (props, ref) {
  var className = props.className,
    children = props.children,
    _props$time = props.time,
    time = _props$time === void 0 ? 3000 : _props$time,
    restProps = _objectWithoutProperties(props, _excluded);
  var _useState = useState(0),
    _useState2 = _slicedToArray(_useState, 2),
    activeIndex = _useState2[0],
    setActiveIndex = _useState2[1];
  var _useState3 = useState(1),
    _useState4 = _slicedToArray(_useState3, 2),
    opacity = _useState4[0],
    setOpacity = _useState4[1];
  var classes = classNames("e-carousel", className);
  useEffect(function () {
    var interval = setInterval(function () {
      setActiveIndex(function (prevIndex) {
        return prevIndex === React.Children.count(children) - 1 ? 0 : prevIndex + 1;
      });
      setOpacity(1); // 重置透明度为1
    }, time);
    return function () {
      return clearInterval(interval);
    };
  }, [children]);
  useEffect(function () {
    if (opacity > 0) {
      var fadeOutTimer = setInterval(function () {
        setOpacity(function (prevOpacity) {
          return prevOpacity - 0.01;
        }); // 透明度每次减少0.1
      }, time / 100); // 将总时间分为10个步骤来递减透明度

      return function () {
        return clearInterval(fadeOutTimer);
      };
    }
  }, [activeIndex]);
  var goToNextSlide = function goToNextSlide() {
    setActiveIndex(function (prevIndex) {
      return prevIndex === React.Children.count(children) - 1 ? 0 : prevIndex + 1;
    });
    setOpacity(1); // 重置透明度为1
  };

  var goToPrevSlide = function goToPrevSlide() {
    setActiveIndex(function (prevIndex) {
      return prevIndex === 0 ? React.Children.count(children) - 1 : prevIndex - 1;
    });
    setOpacity(1); // 重置透明度为1
  };

  return /*#__PURE__*/React.createElement("div", _extends({
    className: classes
  }, restProps), /*#__PURE__*/React.createElement("div", {
    className: "e-carousel-content"
  }, React.Children.map(children, function (child, index) {
    return /*#__PURE__*/React.createElement("div", {
      key: index,
      className: "e-carousel-item ".concat(index === activeIndex ? "e-carousel-item-active" : ""),
      style: {
        opacity: index === activeIndex ? opacity : 0
      }
    }, child);
  })), /*#__PURE__*/React.createElement("div", {
    className: "e-carousel-nav"
  }, React.Children.map(children, function (_, index) {
    return /*#__PURE__*/React.createElement("div", {
      key: index,
      className: "e-carousel-nav-dot ".concat(index === activeIndex ? "e-carousel-nav-dot-active" : ""),
      onClick: function onClick() {
        return setActiveIndex(index);
      }
    });
  })), /*#__PURE__*/React.createElement("div", {
    className: "e-carousel-controls"
  }, /*#__PURE__*/React.createElement("div", {
    className: "e-carousel-control",
    onClick: goToPrevSlide
  }, "<"), /*#__PURE__*/React.createElement("div", {
    className: "e-carousel-control",
    onClick: goToNextSlide
  }, ">")));
});
export default Carousel;