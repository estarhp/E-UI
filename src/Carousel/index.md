---
nav:
  title: components
  path: /components
---

# Carousel

Carousel 轮播图

```jsx
import { Carousel } from 'e-ui';

export default () => {
  return (
    <div style={{ height: '600px', width: '800px' }}>
      <Carousel time={5000}>
        <img
          src="https://s2.loli.net/2023/06/25/ogSYF3GJNOEldsM.jpg"
          alt="Image 1"
        />
        <img
          src="https://s2.loli.net/2023/06/20/7JtaImcMLHC3NbB.jpg"
          alt="Image 2"
        />
        <img
          src="https://s2.loli.net/2023/06/14/8napydHRgx4YAkX.jpg"
          alt="Image 3"
        />
        <img
          src="https://s2.loli.net/2023/06/14/4EhsDYKZl1zLAPu.jpg"
          alt="Image 4"
        />
        <img
          src="https://s2.loli.net/2023/06/10/Ie64zNKECfogmhF.jpg"
          alt="Image 5"
        />
      </Carousel>
    </div>
  );
};
```

| 参数      | 说明           | 类型   | 可选值   | 默认值 |
| --------- | -------------- | ------ | -------- | ------ |
| className | 自定义类名     | string | ---      | ---    |
| time      | 轮播更换的时间 | number | 单位毫秒 | 2000   |
