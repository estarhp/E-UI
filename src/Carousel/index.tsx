import React, {forwardRef, useEffect, useState} from "react";
import classNames from "classnames";
import "./style/index.scss"


export interface CarouselProps {
  className?:string,
  children:React.ReactNode
  time?:number
}

const Carousel = forwardRef<HTMLDivElement,CarouselProps>((props,ref) =>{

  const { className, children,time=3000,...restProps } = props;
  const [activeIndex, setActiveIndex] = useState(0);
  const [opacity, setOpacity] = useState(1);

  const classes = classNames("e-carousel", className);


  useEffect(() => {
    const interval = setInterval(() => {
      setActiveIndex((prevIndex) =>
        prevIndex === React.Children.count(children) - 1 ? 0 : prevIndex + 1

      );
      setOpacity(1); // 重置透明度为1
    }, time);



    return () => clearInterval(interval);

  }, [children]);

  useEffect(() => {
    if (opacity > 0) {
      const fadeOutTimer = setInterval(() => {
        setOpacity((prevOpacity) => prevOpacity - 0.01); // 透明度每次减少0.1
      }, time / 100); // 将总时间分为10个步骤来递减透明度

      return () => clearInterval(fadeOutTimer);
    }
  }, [activeIndex]);


  const goToNextSlide = () => {
    setActiveIndex((prevIndex) =>
      prevIndex === React.Children.count(children) - 1 ? 0 : prevIndex + 1
    );
    setOpacity(1); // 重置透明度为1
  };

  const goToPrevSlide = () => {
    setActiveIndex((prevIndex) =>
      prevIndex === 0 ? React.Children.count(children) - 1 : prevIndex - 1
    );
    setOpacity(1); // 重置透明度为1
  };



  return <div className={classes} {...restProps} ref={ref}>
    <div className="e-carousel-content">
      {
        React.Children.map(
          children,(child,index)=>(
            <div
            key={index}
            className={`e-carousel-item ${
              index === activeIndex ? "e-carousel-item-active" : ""
            }`}
            style={{
              opacity: index === activeIndex ? opacity : 0,
            }}
            >
              {child}
            </div>
          )
        )
      }
    </div>
    <div className="e-carousel-nav">
      {React.Children.map(children, (_, index) => (
        <div
          key={index}
          className={`e-carousel-nav-dot ${
            index === activeIndex ? "e-carousel-nav-dot-active" : ""
          }`}
          onClick={() => setActiveIndex(index)}
        ></div>
      ))}
    </div>
    <div className="e-carousel-controls">
      <div className="e-carousel-control" onClick={goToPrevSlide}>
        {`<`}
      </div>
      <div className="e-carousel-control" onClick={goToNextSlide}>
        {`>`}
      </div>
    </div>
  </div>
});

export default Carousel


