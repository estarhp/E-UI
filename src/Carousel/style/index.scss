.e-carousel {
  width: 100%;
  height: 100%;
  position: relative;
}

.e-carousel-content {
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: space-between;;
  align-items: center;
  flex-wrap: nowrap;

}

.e-carousel-item {
  height: 100%;
  flex-shrink: 0;
  flex-grow: 0;
  width: 100%;
  position: relative;
  display: none;
  opacity: 1;
  transition: opacity linear 1s;
}

.e-carousel-item-active {
  display: inline-block;
  transition: opacity linear 1s;
}

.e-carousel-item img {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%; /* 图片宽度设置为父容器宽度的百分之百 */
  height: 100%; /* 图片高度设置为父容器高度的百分之百 */
  object-fit: fill;
}

.e-carousel-nav {
  cursor: pointer;
  position: absolute;
  bottom: 20px;
  background: transparent;
  width: 40%;
  display: flex;
  align-items: center;
  justify-content: space-evenly;
  transform: translateX(-50%);
  left: 50%;
}

.e-carousel-nav-dot {
   height: 8px;
   width: 8px;
   border-radius: 4px;
   background-color: white;
  opacity: .3;
}

.e-carousel-nav-dot-active{
  background-color: black;
  opacity: .8;
}

.e-carousel-controls {
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  display: flex;
  width: 100%;
  justify-content: space-between;
  font-size: 2rem;
  font-weight: 700;
}

.e-carousel-control {
  text-align: center;
  font-size:2rem;
  opacity: .5;
  border-radius: 1rem;
  cursor: pointer;
  vertical-align: center;
  background-color: transparent;
}






