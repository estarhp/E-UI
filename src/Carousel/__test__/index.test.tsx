import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Carousel, { CarouselProps } from '../index';

describe('Carousel Component', () => {
  let wrapper: ShallowWrapper<CarouselProps>;
  const testChildren = [
    <div key="1">Slide 1</div>,
    <div key="2">Slide 2</div>,
    <div key="3">Slide 3</div>,
  ];

  beforeEach(() => {
    wrapper = shallow<CarouselProps>(<Carousel>{testChildren}</Carousel>);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('renders children correctly', () => {
    for (const child of testChildren) {
      expect(wrapper.contains(child)).toBe(true);
    }
  });

  it('triggers next slide when next control is clicked', () => {
    const nextControl = wrapper.find('.e-carousel-control').at(1);
    nextControl.simulate('click');
    expect(wrapper).toMatchSnapshot();
  });

  it('triggers previous slide when previous control is clicked', () => {
    const prevControl = wrapper.find('.e-carousel-control').at(0);
    prevControl.simulate('click');
    expect(wrapper).toMatchSnapshot();
  });

  // ... (其他测试用例)

  it('applies custom className correctly', () => {
    const testClassName = 'custom-carousel';
    wrapper.setProps({ className: testClassName });

    const carouselDiv = wrapper.find('.e-carousel');
    expect(carouselDiv.hasClass(testClassName)).toBe(true);
  });
});
