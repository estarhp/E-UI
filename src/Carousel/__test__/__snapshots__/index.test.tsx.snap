// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Component triggers next slide when next control is clicked 1`] = `
<div
  className="e-carousel"
>
  <div
    className="e-carousel-content"
  >
    <div
      className="e-carousel-item "
      key="0/.$1"
      style={
        {
          "opacity": 0,
        }
      }
    >
      <div
        key="1"
      >
        Slide 1
      </div>
    </div>
    <div
      className="e-carousel-item e-carousel-item-active"
      key="1/.$2"
      style={
        {
          "opacity": 1,
        }
      }
    >
      <div
        key="2"
      >
        Slide 2
      </div>
    </div>
    <div
      className="e-carousel-item "
      key="2/.$3"
      style={
        {
          "opacity": 0,
        }
      }
    >
      <div
        key="3"
      >
        Slide 3
      </div>
    </div>
  </div>
  <div
    className="e-carousel-nav"
  >
    <div
      className="e-carousel-nav-dot "
      key="0/.$1"
      onClick={[Function]}
    />
    <div
      className="e-carousel-nav-dot e-carousel-nav-dot-active"
      key="1/.$2"
      onClick={[Function]}
    />
    <div
      className="e-carousel-nav-dot "
      key="2/.$3"
      onClick={[Function]}
    />
  </div>
  <div
    className="e-carousel-controls"
  >
    <div
      className="e-carousel-control"
      onClick={[Function]}
    >
      &lt;
    </div>
    <div
      className="e-carousel-control"
      onClick={[Function]}
    >
      &gt;
    </div>
  </div>
</div>
`;

exports[`Carousel Component triggers previous slide when previous control is clicked 1`] = `
<div
  className="e-carousel"
>
  <div
    className="e-carousel-content"
  >
    <div
      className="e-carousel-item "
      key="0/.$1"
      style={
        {
          "opacity": 0,
        }
      }
    >
      <div
        key="1"
      >
        Slide 1
      </div>
    </div>
    <div
      className="e-carousel-item "
      key="1/.$2"
      style={
        {
          "opacity": 0,
        }
      }
    >
      <div
        key="2"
      >
        Slide 2
      </div>
    </div>
    <div
      className="e-carousel-item e-carousel-item-active"
      key="2/.$3"
      style={
        {
          "opacity": 1,
        }
      }
    >
      <div
        key="3"
      >
        Slide 3
      </div>
    </div>
  </div>
  <div
    className="e-carousel-nav"
  >
    <div
      className="e-carousel-nav-dot "
      key="0/.$1"
      onClick={[Function]}
    />
    <div
      className="e-carousel-nav-dot "
      key="1/.$2"
      onClick={[Function]}
    />
    <div
      className="e-carousel-nav-dot e-carousel-nav-dot-active"
      key="2/.$3"
      onClick={[Function]}
    />
  </div>
  <div
    className="e-carousel-controls"
  >
    <div
      className="e-carousel-control"
      onClick={[Function]}
    >
      &lt;
    </div>
    <div
      className="e-carousel-control"
      onClick={[Function]}
    >
      &gt;
    </div>
  </div>
</div>
`;
